File: cpp/components/validation-ctt/HEAD/ctt_testcases/AboutAnnouncementDetails.cpp
/******************************************************************************
* Copyright AllSeen Alliance. All rights reserved.
*
*    Permission to use, copy, modify, and/or distribute this software for any
*    purpose with or without fee is hereby granted, provided that the above
*    copyright notice and this permission notice appear in all copies.
*
*    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
*    WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
*    MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
*    ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
*    WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
*    ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
*    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
******************************************************************************/
#include "stdafx.h"
#include "AboutAnnouncementDetails.h"

#include <alljoyn\AboutKeys.h>

using namespace ajn;
using namespace qcc;

AboutAnnouncementDetails::AboutAnnouncementDetails(AJ_PCSTR t_ServiceName, uint16_t t_Version,
    uint16_t t_Port, const MsgArg& t_ObjectDescription, const MsgArg& t_AboutData) : 
    AboutAnnouncement(t_ServiceName, t_Version, t_Port, t_ObjectDescription, t_AboutData) {}

std::map<String, String> AboutAnnouncementDetails::KEY_SIGNATURE_MAP = 
{
        { AboutKeys::DEVICE_ID, "s" },
        { AboutKeys::DEVICE_NAME, "s" },
        { AboutKeys::DEFAULT_LANGUAGE, "s" },
        { AboutKeys::APP_NAME, "s" },
        { AboutKeys::MANUFACTURER, "s" },
        { AboutKeys::MODEL_NUMBER, "s" },
        { AboutKeys::DESCRIPTION, "s" },
        { AboutKeys::DATE_OF_MANUFACTURE, "s" },
        { AboutKeys::SOFTWARE_VERSION, "s" },
        { AboutKeys::AJ_SOFTWARE_VERSION, "s" },
        { AboutKeys::HARDWARE_VERSION, "s" },
        { AboutKeys::SUPPORT_URL, "s" },
        { AboutKeys::APP_ID, "ay" },
        { AboutKeys::SUPPORTED_LANGUAGES, "as" }
};

bool AboutAnnouncementDetails::checkForCorrectType(String t_Key, String t_Signature)
{
    String expectedSignature = KEY_SIGNATURE_MAP.at(t_Key);

    if ((expectedSignature.c_str() != NULL) && (expectedSignature != t_Signature))
    {
        //throw
    }
    return false;
}

AJ_PSTR AboutAnnouncementDetails::getDeviceId() const
{
    AJ_PSTR deviceId;
    getAboutData()->GetDeviceId(&deviceId);
    return deviceId;
}

uint8_t* AboutAnnouncementDetails::getAppId() const
{
    uint8_t* appId;
    size_t appIdSize;
    getAboutData()->GetAppId(&appId, &appIdSize);
    uint8_t* fixedAppId = new uint8_t[appIdSize];

    for (size_t i = 0; i < appIdSize; ++i)
    {
        fixedAppId[i] = appId[i];
    }

    return fixedAppId;
}

AJ_PSTR AboutAnnouncementDetails::getDeviceName() const
{
    AJ_PSTR deviceName;
    getAboutData()->GetDeviceName(&deviceName);
    return deviceName;
}

AJ_PSTR AboutAnnouncementDetails::getAppName() const
{
    AJ_PSTR appName;
    getAboutData()->GetAppName(&appName);
    return appName;
}

AJ_PSTR AboutAnnouncementDetails::getManufacturer() const
{
    AJ_PSTR manufacturer;
    getAboutData()->GetManufacturer(&manufacturer);
    return manufacturer;
}

AJ_PSTR AboutAnnouncementDetails::getModel() const
{
    AJ_PSTR model;
    getAboutData()->GetModelNumber(&model);
    return model;
}

AJ_PCSTR AboutAnnouncementDetails::getSupportedLanguages() const
{
    AJ_PCSTR supportedLanguages;
    size_t arraySize = getAboutData()->GetSupportedLanguages(&supportedLanguages);
    return supportedLanguages;
}

AJ_PSTR AboutAnnouncementDetails::getDescription() const
{
    AJ_PSTR description;
    getAboutData()->GetDescription(&description);
    return description;
}

AJ_PSTR AboutAnnouncementDetails::getManufactureDate() const
{
    AJ_PSTR manufacturerDate;
    getAboutData()->GetDateOfManufacture(&manufacturerDate);
    return manufacturerDate;
}

AJ_PSTR AboutAnnouncementDetails::getSoftwareVersion() const
{
    AJ_PSTR softwareVersion;
    getAboutData()->GetSoftwareVersion(&softwareVersion);
    return softwareVersion;
}

AJ_PSTR AboutAnnouncementDetails::getAjSoftwareVersion() const
{
    AJ_PSTR ajSoftwareVersion;
    getAboutData()->GetAJSoftwareVersion(&ajSoftwareVersion);
    return ajSoftwareVersion;
}

AJ_PSTR AboutAnnouncementDetails::getHardwareVersion() const
{
    AJ_PSTR hardwareVersion;
    getAboutData()->GetHardwareVersion(&hardwareVersion);
    return hardwareVersion;
}

AJ_PSTR AboutAnnouncementDetails::getSupportUrl() const
{
    AJ_PSTR supportUrl;
    getAboutData()->GetSupportUrl(&supportUrl);
    return supportUrl;
}

bool AboutAnnouncementDetails::supportsInterface(AJ_PSTR t_InterfaceName)
{
    size_t numberOfPaths = getObjectDescriptions()->GetPaths(NULL, 0);
    AJ_PCSTR* paths = new AJ_PCSTR[numberOfPaths];
    getObjectDescriptions()->GetPaths(paths, numberOfPaths);

    for (size_t i = 0; i < numberOfPaths; ++i)
    {
        size_t numberOfInterfaces = getObjectDescriptions()->GetInterfaces(paths[i], NULL, 0);
        AJ_PCSTR* intfs = new AJ_PCSTR[numberOfInterfaces];
        getObjectDescriptions()->GetInterfaces(paths[i], intfs, numberOfInterfaces);

        for (size_t j = 0; j < numberOfInterfaces; ++j)
        {
            if (strcmp(intfs[j], t_InterfaceName) == 0)
            {
                delete[] intfs;
                delete[] paths;
                return true;
            }
        }
        delete[] intfs;
    }
    delete[] paths;

    return false;
}

AJ_PSTR AboutAnnouncementDetails::getDefaultLanguage() const
{
    AJ_PSTR defaultLanguage;
    getAboutData()->GetDefaultLanguage(&defaultLanguage);
    return defaultLanguage;
}

File: java/components/validation-ctt/HEAD/ctt_testcases/src/com/at4wireless/alljoyn/core/about/AboutAnnouncementDetails.java
/*
 *    Copyright (c) Open Connectivity Foundation (OCF), AllJoyn Open Source
 *    Project (AJOSP) Contributors and others.
 *    
 *    SPDX-License-Identifier: Apache-2.0
 *    
 *    All rights reserved. This program and the accompanying materials are
 *    made available under the terms of the Apache License, Version 2.0
 *    which accompanies this distribution, and is available at
 *    http://www.apache.org/licenses/LICENSE-2.0
 *    
 *    Copyright (c) Open Connectivity Foundation and Contributors to AllSeen
 *    Alliance. All rights reserved.
 *    
 *    Permission to use, copy, modify, and/or distribute this software for
 *    any purpose with or without fee is hereby granted, provided that the
 *    above copyright notice and this permission notice appear in all
 *    copies.
 *    
 *    THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL
 *    WARRANTIES WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED
 *    WARRANTIES OF MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE
 *    AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL
 *    DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR
 *    PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
 *    TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
 *    PERFORMANCE OF THIS SOFTWARE.
*/
package com.at4wireless.alljoyn.core.about;

import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.UUID;

import org.alljoyn.about.AboutKeys;
import org.alljoyn.bus.BusException;
import org.alljoyn.bus.Variant;
import org.alljoyn.services.common.BusObjectDescription;
import org.alljoyn.services.common.utils.TransportUtil;


// TODO: Auto-generated Javadoc
/**
 * The Class AboutAnnouncementDetails.
 */
public class AboutAnnouncementDetails extends AboutAnnouncement
{
    
    /** The about data object map. */
    private Map<String, Object> aboutDataObjectMap = null;

    /**
     * Instantiates a new about announcement details.
     *
     * @param serviceName the service name
     * @param port the port
     * @param objectDescriptions the object descriptions
     * @param aboutData the about data
     */
    public AboutAnnouncementDetails(String serviceName, short port, BusObjectDescription[] objectDescriptions, Map<String, Variant> aboutData)
    {
        super(serviceName, port, objectDescriptions, aboutData);
    }

    /**
     * Instantiates a new about announcement details.
     *
     * @param aboutAnnouncement the about announcement
     */
    public AboutAnnouncementDetails(AboutAnnouncement aboutAnnouncement)
    {
        this(aboutAnnouncement.getServiceName(), aboutAnnouncement.getPort(), aboutAnnouncement.getObjectDescriptions(), aboutAnnouncement.getAboutData());
    }

    /**
     * Convert about map.
     *
     * @throws BusException the bus exception
     */
    public void convertAboutMap() throws BusException
    {
        BusException busException = null;
        if (getAboutData() == null)
        {
            throw new BusException("aboutData is null!");
        }
        aboutDataObjectMap = new HashMap<String, Object>(getAboutData().size());
        for (Entry<String, Variant> entry : getAboutData().entrySet())
        {
            try
            {
                String key = entry.getKey();
                Variant variant = entry.getValue();
                String signature = variant.getSignature();
                Object value = null;
                checkForCorrectType(key, signature);
                if ("ay".equals(signature))
                {
                    byte[] byteArray = variant.getObject(new byte[]
                    {}.getClass());
                    value = TransportUtil.byteArrayToUUID(byteArray);
                }
                else if ("as".equals(signature))
                {
                    value = variant.getObject(new String[]
                    {}.getClass());
                }
                else
                {
                    value = variant.getObject(Object.class);
                }
                aboutDataObjectMap.put(key, value);
            }
            catch (BusException e)
            {
                if (busException == null)
                {
                    busException = e;
                }
            }
        }
        if (busException != null)
        {
            throw busException;
        }
    }

    /** The Constant KEY_SIGNATURE_MAP. */
    private static final Map<String, String> KEY_SIGNATURE_MAP = Collections.unmodifiableMap(new HashMap<String, String>()
    {
        private static final long serialVersionUID = 1L;
        {
            put(AboutKeys.ABOUT_DEVICE_ID, "s");
            put(AboutKeys.ABOUT_DEVICE_NAME, "s");
            put(AboutKeys.ABOUT_DEFAULT_LANGUAGE, "s");
            put(AboutKeys.ABOUT_APP_NAME, "s");
            put(AboutKeys.ABOUT_MANUFACTURER, "s");
            put(AboutKeys.ABOUT_MODEL_NUMBER, "s");
            put(AboutKeys.ABOUT_DESCRIPTION, "s");
            put(AboutKeys.ABOUT_DATE_OF_MANUFACTURE, "s");
            put(AboutKeys.ABOUT_SOFTWARE_VERSION, "s");
            put(AboutKeys.ABOUT_AJ_SOFTWARE_VERSION, "s");
            put(AboutKeys.ABOUT_HARDWARE_VERSION, "s");
            put(AboutKeys.ABOUT_SUPPORT_URL, "s");
            put(AboutKeys.ABOUT_APP_ID, "ay");
            put(AboutKeys.ABOUT_SUPPORTED_LANGUAGES, "as");
        }
    });

    /**
     * Check for correct type.
     *
     * @param key the key
     * @param signature the signature
     * @return true, if successful
     * @throws BusException the bus exception
     */
    private boolean checkForCorrectType(String key, String signature) throws BusException
    {
        String expectedSignature = KEY_SIGNATURE_MAP.get(key);
        if ((expectedSignature != null) && (!expectedSignature.equals(signature)))
        {
            throw new BusException(key + " has an incorrect signature: '" + signature + "' expected: '" + expectedSignature + "'");
        }
        return false;
    }

    /**
     * Gets the device id.
     *
     * @return the device id
     */
    public String getDeviceId()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_DEVICE_ID);
    }

    /**
     * Gets the app id.
     *
     * @return the app id
     */
    public UUID getAppId()
    {
        return (UUID) aboutDataObjectMap.get(AboutKeys.ABOUT_APP_ID);
    }

    /**
     * Gets the device name.
     *
     * @return the device name
     */
    public String getDeviceName()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_DEVICE_NAME);
    }

    /**
     * Gets the app name.
     *
     * @return the app name
     */
    public String getAppName()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_APP_NAME);
    }

    /**
     * Gets the manufacturer.
     *
     * @return the manufacturer
     */
    public String getManufacturer()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_MANUFACTURER);
    }

    /**
     * Gets the model.
     *
     * @return the model
     */
    public String getModel()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_MODEL_NUMBER);
    }

    /**
     * Gets the supported languages.
     *
     * @return the supported languages
     */
    public String[] getSupportedLanguages()
    {
        return (String[]) aboutDataObjectMap.get(AboutKeys.ABOUT_SUPPORTED_LANGUAGES);
    }

    /**
     * Gets the description.
     *
     * @return the description
     */
    public String getDescription()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_DESCRIPTION);
    }

    /**
     * Gets the manufacture date.
     *
     * @return the manufacture date
     */
    public String getManufactureDate()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_DATE_OF_MANUFACTURE);
    }

    /**
     * Gets the software version.
     *
     * @return the software version
     */
    public String getSoftwareVersion()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_SOFTWARE_VERSION);
    }

    /**
     * Gets the aj software version.
     *
     * @return the aj software version
     */
    public String getAjSoftwareVersion()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_AJ_SOFTWARE_VERSION);
    }

    /**
     * Gets the hardware version.
     *
     * @return the hardware version
     */
    public String getHardwareVersion()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_HARDWARE_VERSION);
    }

    /**
     * Gets the support url.
     *
     * @return the support url
     */
    public String getSupportUrl()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_SUPPORT_URL);
    }

    /**
     * Supports interface.
     *
     * @param interfaceName the interface name
     * @return true, if successful
     */
    public boolean supportsInterface(String interfaceName)
    {
        for (BusObjectDescription busObjDesc : getObjectDescriptions())
        {
            for (String intf : busObjDesc.interfaces)
            {
                if (intf.equals(interfaceName))
                {
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * Gets the object paths.
     *
     * @param interfaceName the interface name
     * @return the object paths
     */
    public Set<String> getObjectPaths(String interfaceName)
    {
        Set<String> objectPaths = new HashSet<String>();

        for (BusObjectDescription busObjectDescription : getObjectDescriptions())
        {
            for (String intf : busObjectDescription.interfaces)
            {
                if (intf.equals(interfaceName))
                {
                    objectPaths.add(busObjectDescription.getPath());
                }
            }
        }

        return objectPaths;
    }

    /**
     * Gets the default language.
     *
     * @return the default language
     */
    public String getDefaultLanguage()
    {
        return (String) aboutDataObjectMap.get(AboutKeys.ABOUT_DEFAULT_LANGUAGE);
    }
}